File: backend/app/models/puzzle.py
from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class PuzzleMetadata(BaseModel):
    """Metadata for a crossword puzzle."""
    title: Optional[str] = None
    author: Optional[str] = None
    editor: Optional[str] = None
    copyright: Optional[str] = None
    publisher: Optional[str] = None
    date: Optional[str] = None
    dow: Optional[str] = None  # Day of week


class Puzzle(BaseModel):
    """Complete crossword puzzle data model matching NYT JSON format."""

    # Grid dimensions
    size: Dict[str, int] = Field(description="Grid dimensions with 'rows' and 'cols' keys")

    # Grid data
    grid: List[str] = Field(description="Flat array of letters, '.' for black squares")
    gridnums: List[int] = Field(description="Clue numbers for each cell, 0 if no number")

    # Clues
    clues: Dict[str, List[str]] = Field(description="Clues organized by 'across' and 'down'")

    # Answers
    answers: Dict[str, List[str]] = Field(description="Answers organized by 'across' and 'down'")

    # Metadata
    author: Optional[str] = None
    editor: Optional[str] = None
    copyright: Optional[str] = None
    publisher: Optional[str] = None
    date: Optional[str] = None
    dow: Optional[str] = None
    title: Optional[str] = None

    # Additional fields that might be in the JSON
    circles: Optional[List[int]] = None
    shades: Optional[List[int]] = None
    notepad: Optional[str] = None
    jnotes: Optional[str] = None

    @property
    def rows(self) -> int:
        """Get number of rows in the grid."""
        return self.size.get("rows", 0)

    @property
    def cols(self) -> int:
        """Get number of columns in the grid."""
        return self.size.get("cols", 0)

    def validate_grid(self) -> bool:
        """Validate that grid dimensions match the data."""
        expected_length = self.rows * self.cols
        return (
            len(self.grid) == expected_length and
            len(self.gridnums) == expected_length
        )


class PuzzleResponse(BaseModel):
    """Response model for puzzle API endpoints."""
    puzzle: Puzzle
    puzzle_id: str  # Date in YYYY-MM-DD format

File: frontend/src/types/puzzle.ts
// TypeScript types for crossword puzzle data

export interface PuzzleSize {
  rows: number;
  cols: number;
}

export interface Puzzle {
  size: PuzzleSize;
  grid: string[];  // Flat array of letters, '.' for black squares
  gridnums: number[];  // Clue numbers for each cell, 0 if no number
  clues: {
    across: string[];
    down: string[];
  };
  answers: {
    across: string[];
    down: string[];
  };
  author?: string;
  editor?: string;
  copyright?: string;
  publisher?: string;
  date?: string;
  dow?: string;  // Day of week
  title?: string;
  circles?: number[];
  shades?: number[];
  notepad?: string;
  jnotes?: string;
}

export interface PuzzleResponse {
  puzzle: Puzzle;
  puzzle_id: string;
}

// Grid cell with metadata
export interface Cell {
  letter: string;  // Current letter or '.' for black square
  number: number | null;  // Clue number if applicable
  isBlack: boolean;
  row: number;
  col: number;
  isCircled?: boolean;
  isShaded?: boolean;
}

// Clue with cell mapping
export interface ClueInfo {
  number: number;
  clue: string;
  answer: string;
  direction: 'across' | 'down';
  cells: { row: number; col: number }[];  // Cells this clue spans
}

// User's grid state
export interface UserGrid {
  [key: string]: string;  // Key: "row,col", Value: letter
}

// Selected cell/word state
export interface Selection {
  row: number;
  col: number;
  direction: 'across' | 'down';
  clueNumber: number | null;
}

// Validation results
export interface ValidationResult {
  isCorrect: boolean;
  incorrectCells: Set<string>;  // Keys: "row,col"
}
